### PR TITLE
Internal shader modification and custom uniforms

### PR DESCRIPTION
**Pull request by nn4e**
This allows to modify shader code, before it will be sent to the GPU.
Also possible to set custom uniforms.

### CODE REVIEW

File: examples/webgl_internal_shader_modification.html [hunk 1]
```diff
@@ -0,0 +1,298 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - local fogs</title>
```
Comment (WestLangley) [inline examples/webgl_internal_shader_modification.html]: You need to correct the title.

File: examples/webgl_internal_shader_modification.html [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+		<div id="container"></div>
+		<div id="info">
+			<a href="http://threejs.org" target="_blank">three.js</a> webgl - modification of the internal shader<br/>
+		</div>
+		
+
+		<script src="../build/three.js"></script>
+
+		<script src="js/controls/TrackballControls.js"></script>
```
Comment (WestLangley) [inline examples/webgl_internal_shader_modification.html]: Not needed.

File: examples/webgl_internal_shader_modification.html [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+			program.fragmentShader = program.fragmentShader.replace( "//[PrefixBegin]", head );
+			program.fragmentShader = program.fragmentShader.replace("//[PrefixEnd]", fragment_pars );
+			program.fragmentShader = program.fragmentShader.replace( "//[MainEnd]", fragment );
+			//console.log(program.fragmentShader);
+
+		};
+
+		var fogs = [];
+
+		function setupUniforms () {
```
Comment (WestLangley) [inline examples/webgl_internal_shader_modification.html]: Consider renaming to `updateUniforms()` since it is called in the render loop.

File: examples/webgl_internal_shader_modification.html [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+				setupUniforms();
+				//console.log(uniforms.localFogPosition);
+				render();
+				stats.update();
+
+			}
+
+			function render() {
+
+				cameraControls.update();
```
Comment (WestLangley) [inline examples/webgl_internal_shader_modification.html]: `update()` not required in this case.

File: examples/webgl_internal_shader_modification.html [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+				// STATS
+
+				stats = new Stats();
+				container.appendChild( stats.domElement );
+
+				cameraControls = new THREE.OrbitControls(camera, renderer.domElement);
+				cameraControls.target.set( 0, 0, 0);
+				cameraControls.maxDistance = 400;
+				cameraControls.minDistance = 10;
+				cameraControls.update();
```
Comment (WestLangley) [inline examples/webgl_internal_shader_modification.html]: `update()` should not be required here.

**Comment by tschw** [comment]
Why is this change needed? On `dev` uniforms should get properly updated according to `material.uniforms` when `material.needsUpdate`. It used to be broken for such a long time that it may have seemed like intended behavior, though.

It's what the editor does when the uniforms are changed in the JSON. To convince yourself

1. start the editor off the `dev` branch, e.g:
   ```python -m SimpleHTTPServer```
  and open [http://localhost:8000/editor](http://localhost:8000/editor),
2. add a box if there is none by default,
3. change the material to `ShaderMaterial`, and
4. edit JSON and Shaders in random order.

**Comment by nn4e** [comment]
@tschw If you need a functional equivalent PhongMaterial, but with a few additions Shader material forced to do work that can do library

**Comment by tschw** [comment]
Yes, building a `ShaderMaterial` with `.lights`, `.map` etc set, and with the phong code from the shader lib inside seems the way to do it to me. It would be better if that way would became easier instead of adding yet another hack that potentially breaks several subtle assumptions in the renderer.

**Comment by nn4e** [comment]
Btw this has been discussed here https://github.com/mrdoob/three.js/pull/6874

**Comment by tschw** [comment]
Sorry for complaining so late.

Note that I have your very use case, but I can tell from straightening out all the subtle bugs in the renderer for the `ShaderMaterial` case, that this changeset is risky.

It should very well be possible to achieve your goals by making the shader library easier to use for `ShaderMaterial`'s and have the renderer treat them properly where it doesn't do so already. Also cleaner, safer and easier considering maintenance.

@mrdoob Let's have *one* way to get custom code into shaders and not several.

**Comment by tschw** [comment]
The ultimate goal would be to remove the "fixed function" materials once there is no dependent code left in the renderer: At this point they can trivially be constructed from the outside.

**Comment by dubejf** [comment]
Probably relevant #4271

**Comment by WestLangley** [comment]
> Yes, building a ShaderMaterial with .lights, .map etc set, and with the phong code from the shader lib inside seems the way to do it

Agreed. Being able to _easily_ extend `MeshPhongMaterial`  has been a long-awaited feature. Even so, it is doable already.

I agree that would be a better solution than supporting the injection of shader code programatically.

> The ultimate goal would be to remove the "fixed function" materials once there is no dependent code left in the renderer:

Exactly. Having `Phong` etc. as special cases in the renderer is the core of the problem.

@nn4e Please be patient. You will get the capabilities you are looking for eventually. :  - )

**Comment by tschw** [comment]
> @nn4e Please be patient. You will get the capabilities you are looking for eventually. : - )

We better make sure he does. Reading #6874, it appears the ingenious code injection hack was proposed by Ricardo.

> Being able to easily extend MeshPhongMaterial has been a long-awaited feature.
> [...] Having Phong etc. as special cases in the renderer is the core of the problem.

Could implement this feature in a way that promotes the evolution of the renderer rather than adding to its cancer. Maybe like this
```javascript
var shaderBuilder = new ShaderBuilder( MeshPhongSpecification );
shaderBuilder.replaceChunk( DiffuseMapping, proceduralTextureChunk );
shaderBuilder.injectBefore( GammaCorrection, customFogChunk );
shaderBuilder.applyProductTo( aShaderMaterial );
```
, where shader specifications and chunks are properly documented entities...

**Comment by nn4e** [comment]
Also need  to do something with the management of uniforms, now its hardcoded inside render.
They can be moved to material class for example
Uniforms managemen depends on the light management 
light managment also hardcoded in render despite the fact that, the light is in the scene,  not in render.

I have tried to propose a plug-ins system, but in response was proposed this hack

**Comment by tschw** [comment]
Uniforms management takes a different code path for the built-in materials. As said, updating already works properly for `ShaderMaterial` on `dev`.

For lighting set
```javascript
aShaderMaterial.lights = true;
```
and for input textures use e.g.
```javascript
myShaderMaterial.uniforms.map = { type: 't', value: myShaderMaterial.map };
```
could be easier, but it's working.

> I have tried to propose a plug-ins system, but in response was proposed this hack

Yes, I personally believe you were up to something cleaner and we shouldn't ask you to build a hack in order to reject it.

**Comment by mrdoob** [comment]
Would something like this be useful...?

```javascript
var material = new THREE.ShaderMaterial();
material.fromMaterial( new THREE.MeshPhongMaterial );
```

**Comment by liammagee** [comment]
I have been extending MeshLambertMaterial, and was thinking of trying something similar with MeshPhongMaterial. It would be useful, but would it still leave open the problem of how to inject custom chunks (and remove or replace existing chunks)?

**Comment by benaadams** [comment]
@liammagee can you use ShaderMaterial and just add the chunks you want? or RawShaderMaterial if you want to go free form?

**Comment by liammagee** [comment]
@benaadams Thanks, yes that's what I do now (the ShaderMaterial approach). But as someone relatively inexperienced in GLSL, I find the interleaving of chunks and custom code occasionally hard to follow. It also makes derived code more liable to breaking when the shaders are refactored, e.g.  between r70 and r71. I appreciate though these kinds of cases might be edge though, and building another API overlay introduces a maintenance penalty. 

If some of this work is undertaken pre or post-r72, I'd be happy to help - I don't have the knowledge of Three.js internals to undertake it myself.

**Comment by tschw** [comment]
> `shaderMaterial.fromMaterial( new THREE.MeshPhongMaterial() );`
...
> it still leave open the problem of how to inject custom chunks

Yep. But there's no risk in injecting some comments and the overall hack of using those will be a much better one because
* It won't break (possibly subtle) assumptions about the "fixed function" materials in the renderer,
* fixing bugs 

I guess a [*Builder*](http://javarevisited.blogspot.de/2012/06/builder-design-pattern-in-java-example.html) as sketched out above would be the easiest and standard OO way to get c

**Comment by mrdoob** [comment]
> Maybe like this
> ```javascript
var shaderBuilder = new ShaderBuilder( MeshPhongSpecification );
shaderBuilder.replaceChunk( DiffuseMapping, proceduralTextureChunk );
shaderBuilder.injectBefore( GammaCorrection, customFogChunk );
shaderBuilder.applyProductTo( aShaderMaterial );
```

I like the builder idea. As long as we find a better name than "Chunk"... 😄

**Comment by liammagee** [comment]
Re: “chunk” - in the spirit of patterns, maybe Decorator
<http://javarevisited.blogspot.com.au/2011/11/decorator-design-pattern-java-example.html>?
Each chunk in some sense seems to decorate the shader. Could then be:

var shaderBuilder = new ShaderBuilder( MeshPhongSpecification );
shaderBuilder.redecorate( DiffuseMapping, proceduralTextureDecorator ); //
replaceChunk
shaderBuilder.decorate( GammaCorrection, customFogDecorator ); //
injectBefore
shaderBuilder.applyProductTo( aShaderMaterial );

​

**Comment by mrdoob** [comment]
Decorator sounds too Javaish... How about `Part`, `Piece`, `Snippet`, ...?

**Comment by MasterJames** [comment]
"Module"?  
Or maybe Block?

**Comment by WestLangley** [comment]
> As long as we find a better name than "Chunk"...

If a fragment of code is a part of an entire program, then `ShaderFragment` seems reasonable. But that may be confused with fragment shader.

In fact, since we are referring to an object that _contains_ fragments of code, I'd recommend simply `ShaderCode`: 

    ShaderCode[ key ] = value;

**Comment by liammagee** [comment]
Or ShaderProgram? At any rate I'd be happy to help with the implementation
- this looks like it could start as a side-by-side implementation alongside
ShaderLib / ShaderChunk / UniformsLib?

**Comment by MasterJames** [comment]
I'll vote for shaderCode then.

**Comment by tschw** [comment]
OK. I pulled this stuff into a local branch to shape it up for a PR, keeping the contained bugfixes in the shader code and the example, which I will change to use `ShaderMaterial`. Then will take a shot at the builder for simplification in a second step.

> Decorator sounds too Javaish... 

...and commonly means delegation to an associated object of a common supertype, not present in our case.

> How about Part, Piece, Snippet, ...?
> ShaderFragment seems reasonable
> simply ShaderCode

I like "Part" best.  "Code" is very concrete. Maybe someday thingies turn out better to become code generators or some kind of templates... 
Also been thinking about "Fragment" (in the sense of the word - not the OpenGL term, as already mentioned) and "Function" (in the sense of "functionality" - not a syntactic construct), but those have ambiguities.

**Comment by tschw** [comment]
Quick status:

> I will change to use ShaderMaterial. Then will take a shot at the builder for simplification in a second step.

The two-step approach did not work too well - the ported example ended up too far from recommenable practice. I'm still working on a proper solution.

**Comment by mrdoob** [comment]
Thanks for the update!
